### PR TITLE
A module for managing ZFS admin privileges

### PR DESCRIPTION
**Pull request by natefoo**
##### SUMMARY
This module manages [ZFS's administrative privileges](http://docs.oracle.com/cd/E18752_01/html/819-5461/gbchv.html).

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
zfs_delegate_admin

##### ANSIBLE VERSION
```
ansible 2.5.0 (zfs-admin-module 4a6119ce56) last updated 2017/10/24 16:25:41 (GMT -400)
  config file = None
  configured module search path = [u'/home/nate/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/nate/ansible/ansible/lib/ansible
  executable location = /home/nate/ansible/ansible/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```

##### ADDITIONAL INFORMATION
Implements support for the `zfs allow` and `zfs unallow` commands, which delegate ZFS administrative privileges (such as the ability to `send`, `recv`, or `create`) on specific ZFS datasets or filesystems to unprivileged users.

### CODE REVIEW

**Comment by natefoo** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/791

**Comment by xen0l** [comment]
@natefoo you might also be interested in https://github.com/ansible/ansible/pull/19181 (and do additional review).

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: This should be 2.3

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: zfs -> ZFS

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: Does this also work on Linux? If so, it should be added to the list of supported operating systems. It would be good, if you mentioned OpenZFS support as well.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: Please, provide type as it might be helpful. I suppose this is type: str.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: Please, provide type as it might be helpful. I suppose this is type: str.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: Remove empty line.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: get supports default type (https://docs.python.org/2/library/stdtypes.html#dict.get), so you can use _module.prams.get('users', [])_ instead of _module.params.get('users') or []_

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: You are looking for zfs(1M) everytime you run this function. Why not do it in __init__ when initiating class and using a variable, e.g. self._zfs_binary?

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: You are saying that on/off is the accepted values, yet in argument_spec you are using BOOLEANS, which you are converting with module.boolean. Why not accepting booleans here and remove the call to module.boolean()? Also defining type: bool for everyone/local/descendents/recursive would be more consistent with other modules.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (xen0l) [inline lib/ansible/modules/system/zfs_permissions.py]: Missing comment?

**Comment by xen0l** [comment]
@natefoo you might also want to move this module to storage/zfs, so all ZFS related modules are displayed in its own category in the documentation, which helps a lot when referring to documentation.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (tmshn) [inline lib/ansible/modules/system/zfs_permissions.py]: You can use `type='bool'` rather than `choices=BOOLEANS`, with which the module automatically do type cast `module.boolean(module.params.get(...))` for you.
https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1581

(Note that `None` will be passed as is; without conversion. See https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1654-L1655)

Similary, you can use `type='list' default=list()` for `users`, `groups` and `permissions`.
https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1525

**Review by tmshn — CHANGES_REQUESTED** [review]

**Comment by tmshn** [comment]
I don't know if it's required, but some modules have `ANSIBLE_METADATA`.
How about adding this?

E.g.;
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/system/timezone.py#L31-L33

File: lib/ansible/modules/system/zfs_permissions.py
Comment (dagwieers) [reply lib/ansible/modules/system/zfs_permissions.py]: Please don't add `required: false` because that is implicit and unnecessary.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (dagwieers) [inline lib/ansible/modules/system/zfs_permissions.py]: Same. Please remove.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (dagwieers) [inline lib/ansible/modules/system/zfs_permissions.py]: Same. Please remove.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (dagwieers) [inline lib/ansible/modules/system/zfs_permissions.py]: Same. Please remove.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (dagwieers) [inline lib/ansible/modules/system/zfs_permissions.py]: Same. Please remove.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (dagwieers) [inline lib/ansible/modules/system/zfs_permissions.py]: Same. Please remove.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (dagwieers) [inline lib/ansible/modules/system/zfs_permissions.py]: Same. Please remove.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (dagwieers) [inline lib/ansible/modules/system/zfs_permissions.py]: Please move these to the top of the module and only import the stuff you are using from the library.
```python
from ansible.module_utils.basic import AnsibleModule
```

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by natefoo** [comment]
Thanks for the reviews, all. This module was (iirc) based off the zfs module and carried over some out of date practices. I'll incorporate the feedback ASAP.

File: lib/ansible/modules/system/zfs_permissions.py
Comment (senorsmile) [reply lib/ansible/modules/system/zfs_permissions.py]: I disagree.  Many other modules already shipping have required: false all over the place, and I find this very helpful.  More documentation is never a bad thing (imho).

Comment (dagwieers) [reply lib/ansible/modules/system/zfs_permissions.py]: True, older modules do no comply to the new module contribution documentation because it did not exist. However it is now part of this: http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html

BTW It is easier to find which parameters are required, when only the ones that are required have `required: true`. If they all have a `required:` option, you have to find which ones have true on the right line. Not very convenient, hence we have this best practice.

**Comment by dagwieers** [comment]
Please read: http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html

File: lib/ansible/modules/system/zfs_permissions.py
Comment (senorsmile) [reply lib/ansible/modules/system/zfs_permissions.py]: Ah, I hadn't seen those specific recommendations.  Thanks for the link!

File: lib/ansible/modules/system/zfs_permissions.py
Comment (Akasurde) [inline lib/ansible/modules/system/zfs_permissions.py]: use `required_if`

File: lib/ansible/modules/system/zfs_permissions.py
Comment (Akasurde) [inline lib/ansible/modules/system/zfs_permissions.py]: use `required_if`

**Review by Akasurde — DISMISSED** [review]

File: lib/ansible/modules/system/zfs_permissions.py
Comment (natefoo) [reply lib/ansible/modules/system/zfs_permissions.py]: This one wasn't possible because there's not a `required_if_any_of`.

**Comment by natefoo** [comment]
ready_for_review

**Comment by natefoo** [comment]
+affects_2.5
-affects_2.3

**Comment by natefoo** [comment]
Can someone with the privileges fix the labels on this, since @ansibot isn't able to do so?

**Comment by natefoo** [comment]
@dagwieers @Akasurde could you re-review this please?

**Comment by maxamillion** [comment]
@dagwieers @Akasurde re-ping

**Review by dagwieers — APPROVED** [review]

**Comment by dagwieers** [comment]
rebuild_merge

**Comment by dagwieers** [comment]
This one has been in the queue for too long, and only minor changes are needed.
I am going to merge this and fix them in a subsequent commit.

**Comment by mattclay** [comment]
@dagwieers Please don't merge PRs that have not passed tests or are failing tests. If fixes need to be made, they need to be done as part of the PR before merging. If that's not possible, close the original PR and open a new one with the original work plus the necessary fixes.

**Comment by gundalow** [comment]
Examples need updating to include, `- name:` and use `key: value`

**Comment by dagwieers** [comment]
@gundalow This was done in the subsequent PR. The author did not allow me to add commits to this PR.

**Pull request merged**
